src/main/flight/pid.c

```diff
@@ -1168,19 +1115,24 @@ const bool newRcFrame = getShouldUpdateFeedforward();
 
         // no feedforward in launch control
         float feedforwardGain = launchControlActive ? 0.0f : pidRuntime.pidCoefficient[axis].Kf;
-        if (feedforwardGain > 0) {
-            // transition now calculated in feedforward.c when new RC data arrives
-            float feedForward = feedforwardGain * pidSetpointDelta * pidRuntime.pidFrequency;
 
+        if (feedforwardGain > 0) {
+            float feedForward = feedforwardGain * pidSetpointDelta;
 #ifdef USE_FEEDFORWARD
-            pidData[axis].F = shouldApplyFeedforwardLimits(axis) ?
-                applyFeedforwardLimit(axis, feedForward, pidRuntime.pidCoefficient[axis].Kp, currentPidSetpoint) : feedForward;
-#else
-            pidData[axis].F = feedForward;
+            // limit feedforward if rapidly approaching max rate limit for the RP axes
+            const float feedforwardMaxRate = pidRuntime.feedforwardMaxRate[axis];
+            const float Kp = pidRuntime.pidCoefficient[axis].Kp;
+            if (axis < FD_YAW && feedforwardMaxRate != 0.0f) {
+                if (feedForward * currentPidSetpoint > 0.0f) {
+                    if (fabsf(currentPidSetpoint) <= feedforwardMaxRate) {
+                        feedForward = constrainf(feedForward, (-feedforwardMaxRate - currentPidSetpoint) * Kp, (feedforwardMaxRate - currentPidSetpoint) * Kp);
+                    } else {
+                        feedForward = 0;
+                    }
+                }
+            }
```

Comment:
Is this new or old functionality?
It looks a bit out of place compared to all other FF calculations goes pretty far to avoid using the pid rate setpoint.

Reply:
This is the original function, and hasn't been changed in a long time.  I just moved it here from where it was.  It works as it did.

Reply:
I've removed this from pid.c and refactored it within the feedforward() calculation in rc.c.  It ends up a bit simpler there, and doesn't need recalculating every pid loop.